mdtk/downloaders.py

```diff
@@ -0,0 +1,233 @@
+#        warnings.warn(f'Deleting existing directory: {path}')
+        shutil.rmtree(path)
+#        self.downloaded = False
+#        self.extracted = False
+    
+    
+    def download_midi(self, output_path, cache_path=None):
+        """Downloads the MIDI data to output_path"""
+        cache_path = self.cache_path if cache_path is None else cache_path
+        raise NotImplementedError('In order to download MIDI, you must '
+                                  'implement the download_midi method.')
+        
+    
+    def download_csv(self, output_path, cache_path=None):
+        """Downloads the csv data to output_path"""
+        cache_path = self.cache_path if cache_path is None else cache_path
+        raise NotImplementedError('In order to download CSV, you must '
+                                  'implement the download_midi method.')
+        
+        
+
+# TODO: maybe make a base PPDD class and extend this for various options
+class PPDDSept2018Monophonic(DataDownloader):
+    """Patterns for Preditction Development Dataset. Monophonic data only.
+    
+    References
+    ----------
+    https://www.music-ir.org/mirex/wiki/2019:Patterns_for_Prediction
+    """
+    # TODO: add 'sample_size', to allow only a small random sample of the
+    #       total midi files to be copied to the output
+    def __init__(self, cache_path=DEFAULT_CACHE_PATH,
+                 sizes=['small', 'medium', 'large'], clean=False):
+        super().__init__(cache_path = cache_path)
+        self.dataset_name = self.__class__.__name__
+        base_url = ('http://tomcollinsresearch.net/research/data/mirex/'
+                         'ppdd/ppdd-sep2018')
+        self.download_urls = [
+                os.path.join(base_url, f'PPDD-Sep2018_sym_mono_{size}.zip')
+                for size in sizes
+            ]
+        # location of midi files relative to each unzipped directory
+#        self.midi_paths = ['prime_midi', 'cont_true_midi']
+        self.midi_paths = ['prime_midi']
+        self.clean = clean
+        
+    
+    def download_midi(self, output_path, cache_path=None, overwrite=None):
+        # Cleaning paths, and setting up cache dir ============================
+        cache_path = self.cache_path if cache_path is None else cache_path
+        base_path = os.path.join(cache_path, self.dataset_name)
+        make_directory(base_path, overwrite)
+        make_directory(output_path, overwrite)
+        
+        # Downloading the data ================================================
+        zip_paths = []
+        for url in self.download_urls:
+            filename = os.path.basename(url)
+            zip_path = os.path.join(base_path, filename)
+            zip_paths += [zip_path]
+            download_file(url, zip_path, overwrite=overwrite)
+        
+        # Extracting data from zip files ======================================
+        extracted_paths = []
+        for zip_path in zip_paths:
```

Comment:
Why not extract zips directly into output_path? This would speed up processing and avoid a potential bug (below).